micropip/package_index.py

```diff
@@ -0,0 +1,188 @@
+    size: int | None = None
+
+
+@dataclass
+class ProjectInfo:
+    """
+    This class stores common metadata that can be obtained from different APIs (JSON, Simple)
+    provided by PyPI. Responses received from PyPI or other package indexes that support the
+    same APIs must be converted to this class before being processed by micropip.
+    """
+
+    name: str  # Name of the package
+
+    # List of releases available for the package, sorted in ascending order by version.
+    # For each version, list of wheels compatible with the current platform are stored.
+    # If no such wheel is available, the list is empty.
+    releases: dict[Version, Generator[ProjectInfoFile, None, None]]
+
+    @staticmethod
+    def from_json_api(data: dict[str, Any]) -> "ProjectInfo":
+        """
+        Parse JSON API response
+
+        https://warehouse.pypa.io/api-reference/json.html
+        """
+
+        name: str = data.get("info", {}).get("name", "UNKNOWN")
+        releases_raw: dict[str, list[Any]] = data["releases"]
+
+        # Filter out non PEP 440 compliant versions
+        releases: dict[Version, list[Any]] = {}
+        for version_str, fileinfo in releases_raw.items():
+            try:
+                version = Version(version_str)
+                if str(version) != version_str:
+                    continue
+
+            except InvalidVersion:
+                continue
+
+            # Skip empty releases
+            if not fileinfo:
+                continue
+
+            releases[version] = fileinfo
+
+        return ProjectInfo._compatible_only(name, releases)
+
+    @staticmethod
+    def from_simple_api(data: dict[str, Any]) -> "ProjectInfo":
+        """
+        Parse Simple API response
+
+        https://peps.python.org/pep-0503/
+        https://peps.python.org/pep-0691/
+        """
+        name = data["name"]
+
+        # List of versions (PEP 700), this key is not critical to find packages
+        # but it is required to ensure that the same class instance is returned
+        # from JSON and Simple APIs.
+        versions = data.get("versions", [])
+
+        # Group files by version
+        releases: dict[Version, list[Any]] = defaultdict(list)
+
+        for version in versions:
+            if not _is_valid_pep440_version(version):
+                continue
+
+            releases[Version(version)] = []
+
+        for file in data["files"]:
+            filename = file["filename"]
+
+            if not _fast_check_incompatibility(filename):
+                # parsing a wheel filename is expensive, so we do a quick check first
+                continue
+
+            try:
+                version = parse_version(filename)
+            except (InvalidVersion, InvalidWheelFilename):
+                continue
+
+            releases[version].append(file)
+
+        return ProjectInfo._compatible_only(name, releases)
+
+    @staticmethod
+    def _compatible_only(
```

Comment:
This should probably be more a classmethod since it returns an instance of this class?